### PR TITLE
Add TLS support for rdp-enum-encryption

### PR DESCRIPTION
**Pull request by TomSellers**
This PR adds TLS support to `rdp-enum-encryption`. The value that it adds is that it enables determining the RDP protocol version against servers that require TLS and potentially lays the ground work for CredSSP. It also corrects a few values in the RDP payload that were incorrect.

### Windows Server 2016 with TLS required, NLA optional
```shell
PORT     STATE SERVICE       REASON
3389/tcp open  ms-wbt-server syn-ack ttl 128
| rdp-enum-encryption: 
|   Security layer
|     CredSSP (NLA): SUCCESS
|     CredSSP with Early User Auth: SUCCESS
|     Native RDP: FAILED (SSL_REQUIRED_BY_SERVER)
|     RDSTLS: SUCCESS
|     SSL: SUCCESS
|_  RDP Protocol Version:  RDP 10.2 server
```

### Windows Server 2016 with NLA required

NLA is required so we don't see RDP Protocol Version
```shell
3389/tcp open  ms-wbt-server syn-ack ttl 128
| rdp-enum-encryption: 
|   Security layer
|     CredSSP (NLA): SUCCESS
|     CredSSP with Early User Auth: SUCCESS
|     Native RDP: FAILED (HYBRID_REQUIRED_BY_SERVER)
|     RDSTLS: SUCCESS
|_    SSL: FAILED (HYBRID_REQUIRED_BY_SERVER)
```

### Windows Server 2019 with TLS required, NLA optional
```shell
PORT     STATE SERVICE       REASON
3389/tcp open  ms-wbt-server syn-ack ttl 128
| rdp-enum-encryption: 
|   Security layer
|     CredSSP (NLA): SUCCESS
|     CredSSP with Early User Auth: SUCCESS
|     Native RDP: FAILED (SSL_REQUIRED_BY_SERVER)
|     RDSTLS: SUCCESS
|     SSL: SUCCESS
|_  RDP Protocol Version:  RDP 10.6 server
```

### Windows XP
No change in behavior

```shell
PORT     STATE SERVICE       REASON
3389/tcp open  ms-wbt-server syn-ack ttl 128
| rdp-enum-encryption: 
|   Security layer
|     CredSSP (NLA): Unknown
|     CredSSP with Early User Auth: Unknown
|     Native RDP: Unknown
|     RDSTLS: Unknown
|     SSL: Unknown
|   RDP Encryption level: Client Compatible
|     40-bit RC4: SUCCESS
|     56-bit RC4: SUCCESS
|     128-bit RC4: SUCCESS
|     FIPS 140-1: FAILURE
|_  RDP Protocol Version:  RDP 5.x, 6.x, 7.x, or 8.x server
```

### CODE REVIEW

**Comment by TomSellers** [comment]
I looked into  CredSSP.  I can a see the basic auth flow and I've created a script that doesn't fully authenticate but does allow determining host info pre-auth.  It's very similar to other scripts. I'll clean up tonight and submit via a separate PR. 

Rough output from the unfinished script.
```shell
PORT     STATE SERVICE       REASON
3389/tcp open  ms-wbt-server syn-ack ttl 128
| rdp-credssp-info: 
|   Target_Name: W2016
|   NetBIOS_Domain_Name: W2016
|   NetBIOS_Computer_Name: W16GA-SRV01
|   DNS_Domain_Name: W2016.lab
|   DNS_Computer_Name: W16GA-SRV01.W2016.lab
|   DNS_Tree_Name: W2016.lab
|   Product_Version: 10.0.14393
|_  System_Time: 2019-05-30T13:02:39
```

File: nselib/rdp.lua [hunk 1]
```diff
@@ -113,23 +134,58 @@ Packet = {
...
 
-      while data:len() > pos do
-        block_type, block_len  = string.unpack("<I2I2", data, pos)
+      local pos = 3
+
+      local decoder = asn1.ASN1Decoder:new()
+      decoder:registerTagDecoders( tag_decoder )
+
+      local response_result, userdata
+      _, pos = decoder.decodeLength(data, pos)
```
Comment (dmiller-nmap) [inline nselib/rdp.lua]: `_` needs to be local here. You could do `local _, pos = decoder.decodeLength(data, 3)` and eliminate the extra declaration on line 146.

File: scripts/rdp-enum-encryption.nse [hunk 2]
```diff
@@ -96,9 +93,27 @@ local function enum_protocols(host, port)
       -- we can't tell if the protocol is accepted or not.
       table.insert(res_proto, ("%s: Unknown"):format(k))
     end
+
+    -- For servers that require TLS or NLA the only way to get the RDP protocol
+    -- version to negotiate TLS or NLA. This section does that for TLS. There
+    -- is no NLA currently.
+    if status and (v == 1) then
+      status, err = comm.socket:reconnect_ssl()
```
Comment (dmiller-nmap) [inline scripts/rdp-enum-encryption.nse]: `err` needs to be declared local, too. I find things like this with my Lua check script from [the Code Standards page on SecWiki](https://secwiki.org/w/Nmap/Code_Standards#Tools_to_help).

[the remark this answers is not in the document]
Comment (tsellers-r7) [reply scripts/rdp-enum-encryption.nse]: It thought about that.. and then I ran `ssl-cert` and saw that it pulled the certificates without any changes.  I'll revisit that.

[the remark this answers is not in the document]
Comment (dmiller-nmap) [reply scripts/rdp-enum-encryption.nse]: oh, IIRC, it just connects TLS directly without doing preliminary handshaking. That seems to work on some implementations, but I'm not sure how robust it is for all the configurations you're testing.

File: nselib/rdp.lua [hunk 3]
```diff
@@ -183,7 +239,7 @@ Request = {
 
       local data = stdnse.fromhex(
       "7f 65" .. -- BER: Application-Defined Type = APPLICATION 101,
-      "82 01 90" .. -- BER: Type Length = 404 bytes
+      "82 01 94" .. -- BER: Type Length = 404 bytes
```
[the remark this answers is not in the document]
Comment (TomSellers) [reply nselib/rdp.lua]: The value is correct as are most of the others. I'll fixed the one error I found.

File: scripts/rdp-enum-encryption.nse (hunk 2 above)
Comment (TomSellers) [reply scripts/rdp-enum-encryption.nse]: I use `luacheck` but forgot to do so after refactoring.

**Comment by TomSellers** [comment]
@dmiller-nmap 
Feedback addressed. I've retested with Windows Server 2008, 2016, 2019, and Windows XP SP3.

**Pull request closed without merging**